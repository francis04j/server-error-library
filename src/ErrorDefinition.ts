import crypto from 'crypto'

class ErrorDefinition {
    id: string;
    errorId!: string;
  
    statusCode!: number;
  
    message: string;
  
    constructor(statusCode: number, message: string) {
      this.message = message;
      this.statusCode = statusCode;
    } 
  
  
    /**
     * use to attach a unique code to an ErrorDefintion. this can be used by consumers of your error 
     * to reference an error defintion.
     * returned code is unique so you can guarantee distinctiveness  and uniquely identify an error
     * @param code usually a common status code of the error. e.g httpStatusCode
     * @returns an updated ErrorDefinition with an errorCode (unique hash id)
     */
    tag(id: string): ErrorDefinition {
      this.id = id;
      this.errorId = this.generateId(`${this.statusCode}:${this.message}`);
      return this;
    }

    private generateId (id: string) : string { 
      return crypto
      .createHash('MD5')
      .update(id)
      .digest('hex')
      .substr(0, 6)
      .toUpperCase();
    }
  
    toString(): string {
      const message = this.message.replace(/"/g, '""');
  
      return `Error definition "${this.errorId}" [${this.statusCode}]: ${message}`;
    }
  }
  
  export default ErrorDefinition;