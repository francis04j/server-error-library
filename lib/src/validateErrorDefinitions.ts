import flatten from './utils/flatten';
import ErrorDefinition from './ErrorDefinition';

const bail = (definition) => definition instanceof ErrorDefinition;


const validateErrorDefinitions = (errorDefinitions, errorCodeMap = {}) => {
  const flatErrorDefinitions = flatten(errorDefinitions, {bail, separator: ':'});
 //TODO: remove comments:  console.log('errorDefintions', flatErrorDefinitions);
  Object.entries(flatErrorDefinitions).forEach(([id, definition]) => {
   definition.tag(id);

    if (errorCodeMap[definition.errorId]) {
      throw new Error(
        `Duplicate hash for ids ${JSON.stringify(
          errorCodeMap[definition.errorId].id
        )} and ${JSON.stringify(id)}`
      );
    }

    errorCodeMap[definition.errorId] = definition;
  });

  return errorCodeMap;
};


export default validateErrorDefinitions;
