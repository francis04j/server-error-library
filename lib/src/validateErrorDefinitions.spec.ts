import ErrorDefinition from './ErrorDefinition';
import validateErrorDefinitions from './validateErrorDefinitions';

describe('validateErrorDefinitions', () => {
  it('should return error code map', () => {
    const errorDefinitions = {
      LOGIN: {
          AUTH_FAILED: new ErrorDefinition(404, 'User not found or revive failed.'),
          OTHER_FAILED: new ErrorDefinition(15, 'Grpc error.'),
      },
      LOGOUT: new ErrorDefinition(400, 'Bad request: Invalid user'),
    };
   
      const errorCodeMap = validateErrorDefinitions(errorDefinitions);

      expect(errorCodeMap).toBeTruthy();
      expect(Object.keys(errorCodeMap).length).toBe(3);
   
  });

  it('should throw error if Error definitions clashes', () => {
    const collidingErrorDefinitions = {
      LOGIN: {
        AUTH_FAILED: new ErrorDefinition(404, 'User not found or revive failed.'),
        OTHER_FAILED: new ErrorDefinition(404, 'User not found or revive failed.'),
      },
      LOGOUT: new ErrorDefinition(400, 'Bad request: Invalid user'),
    };

    try {
      validateErrorDefinitions(collidingErrorDefinitions)
    }catch(error) {
      expect(error.message).toBe('Duplicate hash for ids "LOGIN:AUTH_FAILED" and "LOGIN:OTHER_FAILED"')
    }
    
    // expect(() =>
    //   validateErrorDefinitions(collidingErrorDefinitions)
    // ).toThrowErrorMatchingSnapshot();

  });
});
