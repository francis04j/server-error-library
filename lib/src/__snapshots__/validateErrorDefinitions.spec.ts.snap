// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateErrorDefinitions should throw error if Error definitions clash 1`] = `"Cannot read properties of undefined (reading 'id')"`;

exports[`validateErrorDefinitions should verify that there are no hash collision in the errorDefinitions 1`] = `"Duplicate hash for ids undefined and \\"HUB:ERROR:ATTORNMENT\\""`;
